raft/src/main/java/org/apache/kafka/raft/KafkaRaftClient.java

```diff
@@ -584,211 +746,236 @@ private void updateVoterConnections(FindQuorumResponseData response) {
         }
     }
 
-    private boolean hasInconsistentLeader(int epoch, OptionalInt leaderId) {
+    private boolean hasConsistentLeader(int epoch, OptionalInt leaderId) {
         // Only elected leaders are sent in the request/response header, so if we have an elected
         // leaderId, it should be consistent with what is in the message.
-        if (epoch != quorum.epoch()) {
-            return false;
+        if (leaderId.isPresent() && leaderId.getAsInt() == quorum.localId) {
+            // The response indicates that we should be the leader, so we verify that is the case
+            return quorum.isLeader();
+        } else if (epoch != quorum.epoch()) {
+            return true;
         } else {
-            if (!quorum.leaderId().isPresent())
-                return false;
             if (!leaderId.isPresent())
```

Comment:
Could we merge the cases into one boolean expression?
```
return epoch != quorum.epoch() || !leaderId.isPresent() || !quorum.leaderId().isPresent() || leaderId.equals(quorum.leaderId());
```

Reply:
We could, but would that be easier to follow? It took me a few iterations to get this check right. At least let me collapse the branches here.

Reply:
Decided to do this after all. Since it is just chaining `||`, I think the suggestion is still clear enough.